pkg/firewall/firewall.go

```diff
@@ -245,14 +256,21 @@ func (soakp *stakeOrActiveKeepPolicy) getKeepAtIndex(
 		return common.HexToAddress(cachedAddress), nil
 	}
 
+	cache.mutex.Lock()
+	defer cache.mutex.Unlock()
+
+	cachedAddress, isCached = cache.address[index.String()]
+	if isCached {
+		return common.HexToAddress(cachedAddress), nil
+	}
```

Comment:
Should we have separate locks for reading?
```suggestion
	cache.mutex.RLock()
	cachedAddress, isCached = cache.address[index.String()]
	cache.mutex.RUnlock()
	if isCached {
		return common.HexToAddress(cachedAddress), nil
	}
	
	cache.mutex.Lock()
	defer cache.mutex.Unlock()
```

Reply:
No, we shouldn't. We need to make sure the second check happens in the critical section with the fetch from the chain. This is just a variation of a double-checked locking pattern.